src/sparseml/exporters/transforms/utils/matching.py

```diff
@@ -0,0 +1,329 @@
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Tuple, Union
+
+from onnx import NodeProto, TensorProto
+
+from sparseml.onnx.utils import ONNXGraph
+
+
+INITIALIZER_MATCH = "__Initializer__"
+"""
+A special tag **ONLY** used for matching code to
+match against initializers in an onnx graph
+"""
+
+
+def optional_node(op_type: str) -> str:
+    """
+    Tells :func:`get_structural_matches` that this op type is an optional one.
+
+    e.g.
+    ```python
+    get_structural_matches(
+        ...,
+        children_ops=[
+            [
+                optional_node("Transpose")
+                "QuantizeLinear",
+            ]
+        ]
+    )
+    ```
+    """
+    return "Optional-" + op_type
+
+
+def get_structural_matches(
+    graph: ONNXGraph,
+    op_type: str,
+    parent_ops: Optional[List[List[str]]] = None,
+    children_ops: Optional[List[List[str]]] = None,
+) -> List["MatchResult"]:
+    """
+    Gathers all nodes in the `graph` that match the `op_type` and the
+    have the specified parent/children structure,
+    controlled via parent_ops/children_ops.
+
+    A simple example just matching against op_type:
+
+        ```python
+        matches = get_structural_matches(graph, op_type="Identity")
+        for match in matches:
+            id_node = match.node
+            assert isinstance(id_node, onnx.NodeProto) and id_node.op_type == "Identity"
+            assert match.parents == []
+            assert match.children == []
+        ```
+
+    `parent_ops` and `children_ops` are list of list of op type strings.
+    It's a list of lists because nodes can have multiple inputs and multiple outputs.
+    So each sub list in `parent_ops` and `children_ops` will be
+    matched against the corresponding entry in node.input/node.output. I.e.
+        - parent_ops[0] will be compared against node.input[0]
+        - parent_ops[2] will be compared against node.input[2]
+        - children_ops[3] will be compared against node.output[3]
+    Because of this, if the length of node.input/node.output is shorter
+    than parent_ops/children_ops, it is not a possible match and will
+    be discarded.
+
+    Here's a simple example of matching against an identity node
+    with a single parent (because there is a single list in parent_ops),
+    and the parent must be an initializer (the INITIALIZER_MATCH value):
+    ```python
+    matches = get_structural_matches(
+        graph,
+        op_type="Identity",
+        parent_ops=[[INITIALIZER_MATCH]]
+    )
+    for match in matches:
+        id_node = match.node
+        (init, ) = match.parents[0]
+        assert isinstance(init, onnx.TensorProto)
+    ```
+
+    Here is a really complicated example with optional nodes and multiple parents. Here
```

Comment:
Can we also add the example with `parent_ops = [[], ["a", "b", ...]]`?

Reply:
Good idea yep